src/rpp/rpp/operators/merge.hpp

```diff
@@ -90,11 +96,11 @@ struct merge_impl
     template<constraint::subscriber_of_type<ValueType> TSub>
     auto operator()(TSub&& subscriber) const
     {
-        auto state = std::make_shared<merge_state>();
+        auto state = std::make_shared<merge_state>(subscriber.get_subscription());
 
         state->count_of_on_completed_needed.fetch_add(1, std::memory_order::relaxed);
 
-        auto subscription = subscriber.get_subscription().make_child();
+        auto subscription = state->childs_subscriptions.make_child();
```

Comment:
Here is the component diagram for subscription from the original behavior.

![image](https://user-images.githubusercontent.com/6766554/188946398-2f07fee7-1234-44b8-986b-af257ede8033.png)


You could get rid of a layer of child subscription since you have child of child subscription here.

Reply:
Oh I see. I plot the subscription tree wrong. It's a brilliant idea and the correct plot should be

```
  { original }                     { new }

sub_a       sub_b          sub_a             sub_b
  \         /                |                 |
   merge_sub                  \  merge_sub_2  / <-- seen by merge_scbr
      |                        \      |      /
      |                          merge_sub_1    <-- used for early unsubscribing
      |                               |
     sub                             sub
```